darkcluster/src/main/java/com/linkedin/darkcluster/impl/DarkClusterStrategyFactoryImpl.java

```diff
@@ -0,0 +1,118 @@
+ * when darkClusterConfig changes are detected, by way of a {@link LoadBalancerClusterListener}
+ */
+public class DarkClusterStrategyFactoryImpl implements DarkClusterStrategyFactory
+{
+  private final ClusterInfoProvider _clusterInfoProvider;
+  private final String _sourceClusterName;
+  private final DarkClusterDispatcher _darkClusterDispatcher;
+  private final Notifier _notifier;
+  private final Map<String, AtomicReference<DarkClusterStrategy>> _darkStrategyMap;
+  private final Random _random;
+  private final LoadBalancerClusterListener _clusterListener;
+  private final DarkClusterVerifierManager _verifierManager;
+
+  public DarkClusterStrategyFactoryImpl(@Nonnull ClusterInfoProvider clusterInfoProvider,
+                                        @Nonnull String clusterName,
+                                        @Nonnull DarkClusterDispatcher darkClusterDispatcher,
+                                        @Nonnull Notifier notifier,
+                                        @Nonnull Random random,
+                                        @Nonnull DarkClusterVerifierManager verifierManager)
+  {
+    _clusterInfoProvider = clusterInfoProvider;
+    _sourceClusterName = clusterName;
+    _notifier = notifier;
+    _darkStrategyMap = new ConcurrentHashMap<>();
+    _random = random;
+    _darkClusterDispatcher = darkClusterDispatcher;
+    _verifierManager = verifierManager;
+    _clusterListener = new LoadBalancerClusterListener()
+    {
+      @Override
+      public void onClusterAdded(String updatedClusterName)
+      {
+        if (_darkStrategyMap.containsKey(updatedClusterName))
+        {
+          // this is a dark cluster name. however, to refresh the strategies, we need to pull the
+          // darkClusterConfigMap on the parent d2 cluster, because that has the properties needed
+          // to recreate the dark cluster strategies, such as the multiplier.
+          try
+          {
+            DarkClusterConfigMap darkConfigMap = _clusterInfoProvider.getDarkClusterConfigMap(_sourceClusterName);
+            for (Map.Entry<String, DarkClusterConfig> entry : darkConfigMap.entrySet())
+            {
+              String darkClusterName = entry.getKey();
+              // just update the cluster that's changed.
+              if (darkClusterName.equals(updatedClusterName))
+              {
+                _darkStrategyMap.put(entry.getKey(), new AtomicReference<>(createStrategy(clusterName, entry.getValue())));
+              }
+            }
+
+          }
+          catch (ServiceUnavailableException e)
+          {
+            _notifier.notify(() -> new RuntimeException("unable to refresh DarkClusterConfigMap for source cluster: " + _sourceClusterName));
+          }
+        }
+      }
+
+      @Override
+      public void onClusterRemoved(String clusterName)
+      {
+        _darkStrategyMap.remove(clusterName);
```

Comment:
Should account for if cluster removed is the source cluster.

Reply:
This should never happen, just like we don't allow clusters to be deleted if there are services in the cluster. In that case, all services need to be removed from the cluster first. Similarly, all dark services and dark clusters need to be removed as well before the source cluster is removed. This code here can't enforce that, the place that enforces that is whatever writes the d2-config. So I won't change anything here.

Reply:
I agree it should not be happening. But I think it would be good to account for it anyway rather than assuming D2 config writer handles it properly since it is an entirely different component. Up to you.

Reply:
there's nothing that can be done here; if I wanted to find the darkClusters that corresponded to the sourceCluster cluster that was being removed, I can't unless I made a duplicate copy outside of the SOT; that would make a big mess to keep in sync. So there's nothing that can be done here for that case, which should never happen. Nor can I even issue a warning, since I wouldn't even know about any dark clusters that correspond to it (only get the clustername, not the property being removed).